docs/book/programs/tracepoints.md

```diff
@@ -1,3 +1,194 @@
+       // read the filename
+       let filename = unsafe {
+            // Get an UTF-8 String from an array of bytes
+            core::str::from_utf8_unchecked(
+                    // Use the address of the kernel's string
+                    // to copy its contents into the array named 'buf'  
+                    match bpf_probe_read_user_str_bytes (
+                        filename_addr as *const u8,
+                        &mut buf,
+                    ) {
+                        Ok(_) =>  &buf,
+                        Err(e)  => {
+                            info!(&ctx, "tracepoint sys_enter_openat called buf_probe failed {}", e);
+                            return Err(e);
+                        }, 
+                    }
+            )
+
+        };
+
+        info!(&ctx, "tracepoint sys_enter_openat called, filename  {}", filename);  
+    }
+    Ok(0)
+}
+```
+
+## eBPF code
+
+- This is the code (its skeleton is generated by `aya-tool`) that runs in the Kernel's eBPF Virtual Machine. The pattern highlights how aya programs are structured: `aya_tracepoint_echo_open(ctx: TracePointContext)` is a public function; it delegates the actual eBPF task to a another function ( `try_aya_tracepoint_echo_open(ctx: TracePointContext) -> Result<u32, i64>`).
+- The `TracePointContext` is one of goodies that Aya brings in. This works as a Rust-aware facade of the internal nuts and bolts that interact with the kernel's own APIs written in 'C'.
+- The name of the file is a `string`  in 'C' (a null-terminated array of `char` s). To access that string, we need to have the address of the byte at the start of the string. The offset at which this address resides, is 24 according to the kernel's documentation (available through the `cat` command mentioned in the code block above). Using the context's `read_at()` function, the address is obtained and held in `filename_addr`. 
+- The content of what `filename_addr` is pointing to, is copied into a byte-array held inside the function, as an UTF-8 string.
+- Print the UTF-8 string.
+
+It works. When run, the output is like this:
+
+```console
+$ RUST_LOG=info cargo xtask run
+[2024-05-16T15:32:29Z INFO  aya_tracepoint_echo_open] Waiting for Ctrl-C...
+[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /proc/meminfo
+[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
+[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
+[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
+```
+
+But the program is not designed correctly. Why? There are two reasons:
+
+1.    The `buf` array is only of 16 bytes. The path-string of files being opened are likely to be much longer than this. The kernel accommodates 4096 bytes of path-string.
```

Comment:
```suggestion
1. The `buf` array is only of 16 bytes. The path-string of files being opened are likely to be much longer than this. The kernel accommodates 4096 bytes of path-string.
```

No need for so many spaces :stuck_out_tongue:

Reply:
Got it!